### ISSUE TITLE
Better codegen for conditional-access + pattern checks.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#69411]
Given code like the following:

```c#
    static void Main(string[] args)
    {
        if (args?.Length is 1)
            Console.WriteLine(true);
    }
```

The compiler currently compiles that similarly to the following:

```c#
int? __t1 = __argsTemp?.Length;
if (__t1.HasValue && __t1.GetValueOrDefault() == 1) ...
```

While correct, this appears to be unnecessarily expensive.  Specifically, because the pattern itself does not have `null` in its acceptance-set, there's no need to actually produce and instantiate a nullable-int on the stack just to test that instead.  Rather, the above would be better compiled to something similar to:

```c#
if (__argsTemp != null && __argsTemp.Length is 1)
```

In other words, once the null check is performed, if it is null, then the value of `args?.Length` would be known to be null, which could never match `1`, so no further checks are necessary.  However, if the null check is performed and the value is not null, then it is known that it is safe to access members of it directly.  This avoids unnecessary nullable allocations on the stack in either case.  

Note: this could also be done for operators like `==` and `!=` on primitive types, when the comparands are constants.

**Comment by jcouv** [comment]
It would be helpful for triage to have some benchmark data (microbenchmark comparing one pattern of usage of System.Nullable to the other).

**Comment by CyrusNajmabadi** [comment]
Great point.  Will add.

**Comment by CyrusNajmabadi** [comment]
Ok.  Wrote this simple bdn test:

```c#
[SimpleJob()]
public class Benchmark
{
    private int[]? _array;

    [Params(
        0,
        1_000,
        10_000)]
    public int Size { get; set; }

    [GlobalSetup]
    public void PrepareArray() => _array = Size == 0 ? null : new int[Size];

    [Benchmark]
    public int ConditionalAccessTest()
    {
        if (_array?.Length > 0)
            return 0;
        else
            return 1;
    }


    [Benchmark]
    public int UnrolledAccess()
    {
        if (_array != null)
        {
            if (_array.Length > 0)
                return 0;
        }

        return 1;
    }
}
```

Results were: 

```
|                Method |  Size |      Mean |     Error |    StdDev |    Median |
|---------------------- |------ |----------:|----------:|----------:|----------:|
| ConditionalAccessTest |     0 | 0.2144 ns | 0.0107 ns | 0.0100 ns | 0.2169 ns |
|        UnrolledAccess |     0 | 0.0100 ns | 0.0185 ns | 0.0164 ns | 0.0027 ns |
| ConditionalAccessTest |  1000 | 0.0128 ns | 0.0124 ns | 0.0104 ns | 0.0076 ns |
|        UnrolledAccess |  1000 | 0.0023 ns | 0.0042 ns | 0.0035 ns | 0.0000 ns |
| ConditionalAccessTest | 10000 | 0.0237 ns | 0.0167 ns | 0.0148 ns | 0.0233 ns |
|        UnrolledAccess | 10000 | 0.0038 ns | 0.0066 ns | 0.0058 ns | 0.0007 ns |
```

This is universally better in all cases.  With the 'null' case, the unrolled vesion is around 20x better.  With the non-null cases it's about 5-6 times better.  

This is a win on CPU and is def better for the stack as well (no need for extra space to store nullables on it).

**Comment by jaredpar** [comment]
@CyrusNajmabadi did you run this benchmark on .NET Core or Framework?

**Comment by canton7** [comment]
Related: https://github.com/dotnet/roslyn/issues/52629. (They don't duplicate, but there's some overlap).

**Comment by alrz** [comment]
Linking to https://github.com/dotnet/roslyn/issues/21962 for switch
